### PR TITLE
BFB-487: Individual Wall Run

### PR DESCRIPTION
**Pull request by m4rtint**
**JIRA Ticket**
https://perigongames.atlassian.net/browse/BFB-487

**Description**
- Created WallRunSO (ScriptableObject)
- Create WallRunData - Pure C#
- Create WallRunContainer - For the Walls
- PlayerWallRun get's the data.

**Screenshots / GIFs**
image
image

### CODE REVIEW

File: Assets/_BForBoss/_Core/Scripts/Environment/WallRunning/WallRunSO.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        [SerializeField]
+        private float _cameraRotateDuration = 1f;
+        [FoldoutGroup("Camera Settings")]
+        [SerializeField] 
+        private float _lookAlongWallRotationSpeed = 3f;
+        [FoldoutGroup("Camera Settings")]
+        [SerializeField] 
+        private float _minLookAlongWallStabilizationAngle = 5f;
+
+        public WallRunData MapToData()
```
Comment (m4rtint) [inline Assets/_BForBoss/_Core/Scripts/Environment/WallRunning/WallRunSO.cs:49]: I'm going to start refactoring scriptable objects in the future. 

Realized we shouldn't be making them implement interfaces, Should just have a "Domain" object - pure data, and a Scriptable object - Designer facing for value changes. 

So for testing/tools, you just need to create a pure data "domain" object and pass it into whatever needs it. Act's the same way as an interface, i just used an interface incorrectly.

File: Assets/_BForBoss/_Core/Scripts/Character/PlayerWallRunBehaviour.cs [hunk 2]
```diff
@@ -404,9 +369,9 @@ private float GetCameraRoll()
             float targetAngle = 0;
             if (wallDirection != 0)
             {
-                targetAngle = Mathf.Sign(wallDirection) * heading * _maxCameraAngleRoll;
+                targetAngle = Mathf.Sign(wallDirection) * heading * _wallRunData.MaxCameraAngleRoll;
             }
-            return Mathf.LerpAngle(cameraAngle, targetAngle, Mathf.Max(_timeSinceWallAttach, _timeSinceWallDetach) / _cameraRotateDuration);
+            return Mathf.LerpAngle(cameraAngle, targetAngle, Mathf.Max(_timeSinceWallAttach, _timeSinceWallDetach) / _wallRunData.CameraRotateDuration);
```
Comment (Gursaanj) [inline Assets/_BForBoss/_Core/Scripts/Character/PlayerWallRunBehaviour.cs:374]: For safety reasons, we should check to see if `_wallRunData.CameraRotateDuration` is 0 first (and if that is the case then divide by 1). 

Alternatively, we can add a min value for `CameraRotateDuration`. Would still prefer if we do the check as well.

File: Assets/_BForBoss/_Core/Scripts/Environment/WallRunning/WallRunSO.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+        [FoldoutGroup("Camera Settings")]
+        [SerializeField]
+        private float _maxCameraAngleRoll = 30f;
+        [FoldoutGroup("Camera Settings")]
+        [SerializeField]
+        private float _cameraRotateDuration = 1f;
+        [FoldoutGroup("Camera Settings")]
+        [SerializeField] 
+        private float _lookAlongWallRotationSpeed = 3f;
```
Comment (Gursaanj) [inline Assets/_BForBoss/_Core/Scripts/Environment/WallRunning/WallRunSO.cs:44]: I think this field and `_minLookAlongWallStabilizationAngle` should have tooltips as well

File: Assets/_BForBoss/_Core/Scripts/Character/PlayerWallRunBehaviour.cs (hunk 2 above)
Comment (m4rtint) [reply Assets/_BForBoss/_Core/Scripts/Character/PlayerWallRunBehaviour.cs:374]: So default values in struct can only be the built in C# default value which is 0. 

So i'm setting up a Max(cameraiValue, 0.01)

File: Assets/_BForBoss/_Core/Scripts/Environment/WallRunning/WallRunSO.cs (hunk 3 above)
Comment (m4rtint) [reply Assets/_BForBoss/_Core/Scripts/Environment/WallRunning/WallRunSO.cs:45]: Updated

**Review by Gursaanj — APPROVED** [review]

**Pull request merged**